File: python/inference.py
"""
Hand position inference from fingered notes.

Reads a JSON file of fingered notes and predicts hand positions.
"""

import argparse
import json
from pathlib import Path
import numpy as np
import torch

from model import HandPositionTransformer

PROJECT_ROOT = Path(__file__).resolve().parent.parent

BLACK_KEY_NOTES = [1, 4, 6, 9, 11]


def is_black_key(midi: int) -> float:
    if midi < 0:
        return -1.0
    return 1.0 if ((midi - 21) % 12) in BLACK_KEY_NOTES else 0.0


def load_model(hand: str):
    """Load trained hand position transformer."""
    model_path = PROJECT_ROOT / "checkpoints" / f"hand_position_{hand}.pt"
    checkpoint = torch.load(model_path, map_location='cpu', weights_only=True)

    model = HandPositionTransformer(
        d_model=checkpoint['d_model'],
        nhead=checkpoint['nhead'],
        num_layers=checkpoint['num_layers'],
        dim_feedforward=checkpoint.get('dim_feedforward', 128)
    )
    model.load_state_dict(checkpoint['model_state'])
    model.eval()
    return model


def group_notes_by_time(notes: list[dict], threshold_ms: float = 1.0) -> list[list[dict]]:
    """Group simultaneous notes into chords."""
    if not notes:
        return []

    notes = sorted(notes, key=lambda n: n['time'])
    groups = []
    current_group = [notes[0]]

    for note in notes[1:]:
        if note['time'] - current_group[0]['time'] <= threshold_ms:
            current_group.append(note)
        else:
            groups.append(current_group)
            current_group = [note]

    groups.append(current_group)
    return groups


def build_tokens(prev_hand: dict, current_notes: list[dict], lookahead: list[dict]) -> np.ndarray:
    """
    Build 21x4 input tokens for hand position prediction.

    Args:
        prev_hand: dict with 'wrist': (x,y) and 'fingertips': [(x,y), ...]
        current_notes: list of notes with 'note' (midi) and 'finger' (1-5)
        lookahead: list of future notes with 'note', 'finger', and 'time_until'
    """
    tokens = np.zeros((21, 4), dtype=np.float32)

    # --- Previous hand tokens (0-5) ---
    tokens[0, 0] = prev_hand['wrist'][0]
    tokens[0, 1] = prev_hand['wrist'][1]
    tokens[0, 2] = 0.0  # part_id
    tokens[0, 3] = 0.0  # token_type

    for f in range(5):
        tokens[1 + f, 0] = prev_hand['fingertips'][f][0]
        tokens[1 + f, 1] = prev_hand['fingertips'][f][1]
        tokens[1 + f, 2] = (f + 1) / 5.0
        tokens[1 + f, 3] = 0.0

    # --- Current note tokens (6-10): one per finger slot ---
    finger_to_midi = {}
    for note in current_notes:
        finger_to_midi[note['finger'] - 1] = note['note']

    for f in range(5):
        if f in finger_to_midi:
            midi = finger_to_midi[f]
            tokens[6 + f, 0] = (midi - 21) / 87.0
            tokens[6 + f, 1] = is_black_key(midi)
            tokens[6 + f, 2] = 1.0  # is_active
        else:
            tokens[6 + f, 0] = -1.0
            tokens[6 + f, 1] = -1.0
            tokens[6 + f, 2] = -1.0
        tokens[6 + f, 3] = 0.33

    # --- Lookahead tokens (11-20) ---
    for k in range(10):
        if k < len(lookahead):
            note = lookahead[k]
            tokens[11 + k, 0] = (note['note'] - 21) / 87.0
            tokens[11 + k, 1] = is_black_key(note['note'])
            tokens[11 + k, 2] = (note['finger'] - 1) / 4.0
            tokens[11 + k, 3] = min(note['time_until'], 10.0) / 10.0
        else:
            tokens[11 + k, :] = -1.0

    return tokens


def predict(model, prev_hand: dict, current_notes: list[dict], lookahead: list[dict]) -> dict:
    """Predict hand position for current notes."""
    tokens = build_tokens(prev_hand, current_notes, lookahead)

    with torch.no_grad():
        tokens_tensor = torch.tensor([tokens], dtype=torch.float32)
        output = model(tokens_tensor).numpy()[0]

    return {
        'wrist': (float(output[0]), float(output[1])),
        'fingertips': [
            (float(output[2]), float(output[3])),
            (float(output[4]), float(output[5])),
            (float(output[6]), float(output[7])),
            (float(output[8]), float(output[9])),
            (float(output[10]), float(output[11])),
        ]
    }


def predict_hand_positions(notes: list[dict], hand: str, model) -> list[dict]:
    """Predict hand positions for all chords of one hand."""
    is_left = (hand == 'left')
    hand_notes = [n for n in notes if n['left'] == is_left]

    if not hand_notes:
        return []

    groups = group_notes_by_time(hand_notes)

    # Initialize hand position at center
    prev_hand = {
        'wrist': (0.5, 1.5),
        'fingertips': [
            (0.46, 0.8),
            (0.48, 0.6),
            (0.50, 0.5),
            (0.52, 0.6),
            (0.54, 0.8),
        ]
    }

    results = []

    for i, group in enumerate(groups):
        current_time = group[0]['time']

        # Build lookahead from future groups
        lookahead = []
        for future_group in groups[i + 1:]:
            future_time = future_group[0]['time']
            time_until = (future_time - current_time) / 1000.0

            for note in future_group:
                lookahead.append({
                    'note': note['note'],
                    'finger': note['finger'],
                    'time_until': time_until
                })
                if len(lookahead) >= 10:
                    break
            if len(lookahead) >= 10:
                break

        # Predict
        hand_pos = predict(model, prev_hand, group, lookahead)

        results.append({
            'time': current_time,
            'notes': [{'note': n['note'], 'finger': n['finger']} for n in group],
            'wrist': hand_pos['wrist'],
            'fingertips': hand_pos['fingertips']
        })

        prev_hand = hand_pos

    return results


def main():
    parser = argparse.ArgumentParser(description="Hand position inference from fingered notes")
    parser.add_argument("input", help="Input JSON file with fingered notes")
    parser.add_argument("-o", "--output", help="Output JSON file (default: input_positions.json)")
    parser.add_argument("--hand", "-H", choices=["left", "right", "both"], default="both",
                        help="Which hand to predict (default: both)")

    args = parser.parse_args()

    # Load input
    with open(args.input) as f:
        notes = json.load(f)

    print(f"Loaded {len(notes)} notes")

    results = {'left': [], 'right': []}

    hands = ['left', 'right'] if args.hand == 'both' else [args.hand]

    for hand in hands:
        model = load_model(hand)
        print(f"Loaded {hand} hand model")
        results[hand] = predict_hand_positions(notes, hand, model)
        print(f"Predicted {len(results[hand])} positions for {hand} hand")

    # Save output
    output_path = args.output or args.input.replace('.json', '_positions.json')
    with open(output_path, 'w') as f:
        json.dump(results, f, indent=2)

    print(f"Saved: {output_path}")


if __name__ == "__main__":
    main()

File: js/src/inference.ts
import { InferenceSession, Tensor } from "onnxruntime-common";

const BLACK_KEY_NOTES = new Set([1, 4, 6, 9, 11]); // A#, C#, D#, F#, G#

function isBlackKey(midi: number): number {
  if (midi < 0) return -1.0;
  const keyIndex = midi - 21;
  return BLACK_KEY_NOTES.has(((keyIndex % 12) + 12) % 12) ? 1.0 : 0.0;
}

export interface HandPosition {
  wrist: [number, number];
  fingertips: [[number, number], [number, number], [number, number], [number, number], [number, number]];
}

export interface Note {
  left: boolean;
  note: number;
  time: number;
  duration: number;
  finger: number; // 1-5
  velocity?: number;
}

interface LookaheadNote {
  note: number;
  finger: number;
  timeUntil: number;
}

/**
 * Build 21x4 input tokens for hand position prediction.
 * Returns a flat Float32Array of 84 values in row-major order.
 */
function buildTokens(
  prevHand: HandPosition,
  currentNotes: { note: number; finger: number }[],
  lookahead: LookaheadNote[]
): Float32Array {
  const tokens = new Float32Array(21 * 4);

  // --- Previous hand tokens (0-5) ---
  // Wrist (token 0)
  tokens[0 * 4 + 0] = prevHand.wrist[0];
  tokens[0 * 4 + 1] = prevHand.wrist[1];
  tokens[0 * 4 + 2] = 0.0; // part_id
  tokens[0 * 4 + 3] = 0.0; // token_type

  // Fingertips (tokens 1-5)
  for (let f = 0; f < 5; f++) {
    const offset = (1 + f) * 4;
    tokens[offset + 0] = prevHand.fingertips[f][0];
    tokens[offset + 1] = prevHand.fingertips[f][1];
    tokens[offset + 2] = (f + 1) / 5.0;
    tokens[offset + 3] = 0.0;
  }

  // --- Current note tokens (6-10): one per finger slot ---
  const fingerToMidi = new Map<number, number>();
  for (const note of currentNotes) {
    fingerToMidi.set(note.finger - 1, note.note);
  }

  for (let f = 0; f < 5; f++) {
    const offset = (6 + f) * 4;
    const midi = fingerToMidi.get(f);

    if (midi !== undefined) {
      tokens[offset + 0] = (midi - 21) / 87.0;
      tokens[offset + 1] = isBlackKey(midi);
      tokens[offset + 2] = 1.0; // is_active
    } else {
      tokens[offset + 0] = -1.0;
      tokens[offset + 1] = -1.0;
      tokens[offset + 2] = -1.0;
    }
    tokens[offset + 3] = 0.33;
  }

  // --- Lookahead tokens (11-20) ---
  for (let k = 0; k < 10; k++) {
    const offset = (11 + k) * 4;

    if (k < lookahead.length) {
      const ln = lookahead[k];
      tokens[offset + 0] = (ln.note - 21) / 87.0;
      tokens[offset + 1] = isBlackKey(ln.note);
      tokens[offset + 2] = (ln.finger - 1) / 4.0;
      tokens[offset + 3] = Math.min(ln.timeUntil, 10.0) / 10.0;
    } else {
      tokens[offset + 0] = -1.0;
      tokens[offset + 1] = -1.0;
      tokens[offset + 2] = -1.0;
      tokens[offset + 3] = -1.0;
    }
  }

  return tokens;
}

function parseOutput(data: Float32Array): HandPosition {
  return {
    wrist: [data[0], data[1]],
    fingertips: [
      [data[2], data[3]],
      [data[4], data[5]],
      [data[6], data[7]],
      [data[8], data[9]],
      [data[10], data[11]],
    ],
  };
}

interface NoteGroup {
  time: number;
  notes: { note: number; finger: number }[];
}

function groupNotesByTime(notes: Note[], thresholdMs = 1.0): NoteGroup[] {
  if (notes.length === 0) return [];

  const sorted = [...notes].sort((a, b) => a.time - b.time);
  const groups: NoteGroup[] = [];
  let currentGroup: NoteGroup = { time: sorted[0].time, notes: [] };

  for (const note of sorted) {
    if (note.time - currentGroup.time <= thresholdMs) {
      currentGroup.notes.push({ note: note.note, finger: note.finger });
    } else {
      groups.push(currentGroup);
      currentGroup = { time: note.time, notes: [{ note: note.note, finger: note.finger }] };
    }
  }
  groups.push(currentGroup);

  return groups;
}

export interface PositionResult {
  time: number;
  notes: { note: number; finger: number }[];
  position: HandPosition;
}

/**
 * Predict hand positions for all chords of one hand.
 */
export async function predictHandPositions(
  notes: Note[],
  isLeft: boolean,
  model: InferenceSession
): Promise<PositionResult[]> {
  const handNotes = notes.filter((n) => n.left === isLeft);

  if (handNotes.length === 0) return [];

  const groups = groupNotesByTime(handNotes);

  // Initialize hand position at center
  let prevHand: HandPosition = {
    wrist: [0.5, 1.5],
    fingertips: [
      [0.46, 0.8],
      [0.48, 0.6],
      [0.5, 0.5],
      [0.52, 0.6],
      [0.54, 0.8],
    ],
  };

  const results: PositionResult[] = [];

  for (let i = 0; i < groups.length; i++) {
    const group = groups[i];
    const currentTime = group.time;

    // Build lookahead from future groups
    const lookahead: LookaheadNote[] = [];
    for (let j = i + 1; j < groups.length && lookahead.length < 10; j++) {
      const futureGroup = groups[j];
      const timeUntil = (futureGroup.time - currentTime) / 1000.0;

      for (const note of futureGroup.notes) {
        lookahead.push({
          note: note.note,
          finger: note.finger,
          timeUntil,
        });
        if (lookahead.length >= 10) break;
      }
    }

    // Build tokens and run model
    const tokenData = buildTokens(prevHand, group.notes, lookahead);
    const inputTensor = new Tensor("float32", tokenData, [1, 21, 4]);
    const output = await model.run({ tokens: inputTensor });
    const positions = output.positions.data as Float32Array;
    const handPos = parseOutput(positions);

    results.push({
      time: currentTime,
      notes: group.notes,
      position: handPos,
    });

    prevHand = handPos;
  }

  return results;
}

export type Models = { left: InferenceSession; right: InferenceSession };

export interface PredictionResult {
  left: PositionResult[];
  right: PositionResult[];
}

/**
 * Predict hand positions for a sequence of fingered piano notes.
 *
 * @param notes - Array of notes with finger assignments (1-5)
 * @param models - Pre-loaded ONNX inference sessions for left and right hand models
 * @returns Hand positions for each chord, grouped by hand
 */
export async function predictPositions(
  notes: Note[],
  models: Models
): Promise<PredictionResult> {
  const [leftResults, rightResults] = await Promise.all([
    predictHandPositions(notes, true, models.left),
    predictHandPositions(notes, false, models.right),
  ]);

  return {
    left: leftResults,
    right: rightResults,
  };
}
